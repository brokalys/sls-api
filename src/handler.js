import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
