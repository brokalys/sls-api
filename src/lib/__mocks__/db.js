export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
