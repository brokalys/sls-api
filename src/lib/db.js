import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
