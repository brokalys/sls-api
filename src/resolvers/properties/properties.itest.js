import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import db from 'lib/db';

jest.mock('lib/db');
jest.mock('lib/sql-cache');

describe('Query: properties', () => {
  let query;

  beforeEach(() => {
    process.env.BROKALYS_PRIVATE_KEY = 'PRIVATE_KEY';

    const utils = createTestClient(server);
    query = utils.query;
  });

  afterEach(jest.resetAllMocks);

  test('successfully retrieves property summary.count data', async () => {
    db.query.mockReturnValueOnce([{ count: 120 }]);

    const response = await query({
      query: `
        {
          properties {
            summary {
              count
            }
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves property summary.price data', async () => {
    db.query.mockReturnValueOnce([{ price: 100 }, { price: 200 }]);

    const response = await query({
      query: `
        {
          properties {
            summary {
              price {
                median
              }
            }
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  // @todo: passing headers to integration tests is currently not possible..
  // @see https://github.com/apollographql/apollo-server/issues/2277
  xtest('successfully retrieves results', async () => {
    db.query.mockReturnValueOnce([{ id: 1 }, { id: 2 }]);

    const response = await query({
      query: `
        {
          properties {
            results {
              id
            }
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test.each([
    { filter: { published_at: { gte: 'value' } } }, // simply wrong value
    { filter: { published_at: '2019-01-01' } }, // missing filter expression
  ])(
    'throws a validation error when trying to retrieve summary.count',
    async (input) => {
      db.query.mockReturnValueOnce([{ count: 120 }]);

      const response = await query({
        query: `
        query GetCount($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              count
            }
          }
        }
      `,
        variables: input,
      });

      expect(response.errors).toHaveLength(1);
    },
  );

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    db.query.mockReturnValueOnce([{ id: 1 }, { id: 2 }]);

    const response = await query({
      query: `
        {
          properties {
            results {
              id
            }
          }
        }
      `,
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });
});
