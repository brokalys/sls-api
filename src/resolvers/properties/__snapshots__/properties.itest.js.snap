// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query: properties property handles internal errors 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": null,
    },
  },
  "errors": Array [
    Object {
      "message": "An unexpected error occurred. Pleas try again later.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties returns 0 count when no properties are found 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "count": 0,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves property summary.count data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "count": 120,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves property summary.price data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "price": Object {
          "count": 2,
          "max": 200,
          "mean": 150,
          "median": 150,
          "min": 100,
          "mode": 100,
          "standardDev": 50,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves results 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves truncated property summary.price data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "price": Object {
          "count": 2,
          "max": 300,
          "mean": 250,
          "median": 250,
          "min": 200,
          "mode": 200,
          "standardDev": 50,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
